spot_driver_cpp/src/interfaces/spot_interface.cpp

```diff
@@ -0,0 +1,394 @@
+  {
+    return builtin_interfaces::build<builtin_interfaces::msg::Time>().sec(seconds_unskewed).nanosec(nanos_unskewed);
+  }
+}
+
+tl::expected<sensor_msgs::msg::CameraInfo, std::string> toCameraInfoMsg(const bosdyn::api::ImageResponse& image_response, const std::string& robot_name, const google::protobuf::Duration& clock_skew)
+{
+  sensor_msgs::msg::CameraInfo info_msg;
+  info_msg.distortion_model = sensor_msgs::distortion_models::PLUMB_BOB;
+  info_msg.height = image_response.shot().image().rows();
+  info_msg.width = image_response.shot().image().cols();
+  info_msg.header.frame_id = robot_name + "/" + image_response.shot().frame_name_image_sensor();
+  info_msg.header.stamp = applyClockSkew(image_response.shot().acquisition_time(), clock_skew);
+
+  // We assume that the camera images have already been corrected for distortion, so the 5 distortion parameters are all zero
+  info_msg.d = std::vector<double>{5, 0.0};
+
+  // Set the rectification matrix to identity, since this is not a stereo pair.
+  info_msg.r[0] = 1.0;
+  info_msg.r[1] = 0.0;
+  info_msg.r[2] = 0.0;
+  info_msg.r[3] = 0.0;
+  info_msg.r[4] = 1.0;
+  info_msg.r[5] = 0.0;
+  info_msg.r[6] = 0.0;
+  info_msg.r[7] = 0.0;
+  info_msg.r[8] = 1.0;
+
+  const auto& intrinsics = image_response.source().pinhole().intrinsics();
+
+  // Create the 3x3 intrinsics matrix.
+  info_msg.k[0] = intrinsics.focal_length().x();
+  info_msg.k[2] = intrinsics.principal_point().x();
+  info_msg.k[4] = intrinsics.focal_length().y();
+  info_msg.k[5] = intrinsics.principal_point().y();
+  info_msg.k[8] = 1.0;
+
+  // All Spot cameras are functionally monocular, so Tx and Ty are not set here.
+  info_msg.p[0] = intrinsics.focal_length().x();
+  info_msg.p[2] = intrinsics.principal_point().x();
+  info_msg.p[5] = intrinsics.focal_length().y();
+  info_msg.p[6] = intrinsics.principal_point().y();
+  info_msg.p[10] = 1.0;
+
+  return info_msg;
+}
+
+tl::expected<sensor_msgs::msg::Image, std::string> toImageMsg(const bosdyn::api::ImageCapture& image_capture, const std::string& robot_name, const google::protobuf::Duration& clock_skew)
+{
+      const auto& image = image_capture.image();
+      auto data = image.data();
+
+      std_msgs::msg::Header header;
+      header.frame_id = robot_name + "/" + image_capture.frame_name_image_sensor();
+      header.stamp = applyClockSkew(image_capture.acquisition_time(), clock_skew);
+
+      const auto pixel_format_cv = getCvPixelFormat(image.pixel_format());
+      if (!pixel_format_cv)
+      {
+        return tl::make_unexpected("Failed to determine pixel format: " + pixel_format_cv.error());
+      }
+
+      if (image.format() == bosdyn::api::Image_Format_FORMAT_JPEG)
+      {
+        // When the image is JPEG-compressed, it is represented as a 1 x (width * height) row of bytes.
+        // First we create a cv::Mat which contains the compressed image data...
+        const cv::Mat img_compressed{ 1, image.rows() * image.cols(), CV_8UC1, &data.front()};
```

Comment:
This would be a good thing to investigate and improve as part of a future PR. I think as currently implemented this is adequately performant (especially compared to the performance of the current Python-based image publisher, oof), but there's probably an opportunity to squeeze some extra efficiency out of this too.